### PR TITLE
Add notJwt, maxTokenExpirySeconds, and skipTypeCheck to CRDs

### PR DESCRIPTION
**Pull request by mstruk**
Signed-off-by: Marko Strukelj <marko.strukelj@gmail.com>

This is a preliminary PR, not yet ready to merge. It is dependent on `strimzi-kafka-oauth` 0.2.0 being released. For the time being it uses 1.0.0-SNAPSHOT version for `strimzi-kafka-oauth` - meaning that it requires `strimzi-kafka-oauth` being built locally from master branch.

Some checklist items may still need to be addressed.

### Type of change

- Enhancement / new feature

### Description

`strimzi-kafka-oauth` module introduced additional oauth authentication config options which need to be exposed through CRDs.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by scholzj** [comment]
@mstruk Could you update this against the 0.2.0 version of the OAuth library so that we can run the tests?

**Comment by tombentley** [comment]
@mstruk is this ready for review yet?

**Comment by scholzj** [comment]
@mstruk Culd oyu update this to OAuth 0.2.0 so that it is building? We can than review it and merge it.

**Comment by mstruk** [comment]
@scholzj @tombentley Sorry about not responding earlier to this. Bad gmail filter siphoned the notification emails out of my Inbox :O 

I think this PR is ready to merge. I bumped strimzi-kafka-oauth version to 0.2.0 and fixed some docs and tests issues, added some more tests. 

CI doesn't pass but I don't see how the failure would be connected to this PR. I haven't tried manually setup the cluster with the built images for final test, but I don't see why that wouldn't work considering the additional tests pass fine.

**Comment by scholzj** [comment]
@mstruk It looks like you didn't updated the CRDs for the Helm Chart:

```
ERROR: Uncommitted changes in derived resources:
M	helm-charts/strimzi-kafka-operator/templates/041-Crd-kafkaconnect.yaml
M	helm-charts/strimzi-kafka-operator/templates/042-Crd-kafkaconnects2i.yaml
M	helm-charts/strimzi-kafka-operator/templates/046-Crd-kafkabridge.yaml
Run the following to add up-to-date resources:
  mvn clean verify -DskipTests -DskipITs \
    && make crd_install \
    && git add install/ helm-charts/ documentation/modules/appendix_crds.adoc cluster-operator/src/main/resources/cluster-roles
    && git commit -s -m 'Update derived resources'
```

Could you please run `make crd_install` and commit the changes?

File: api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java [hunk 1]
```diff
@@ -114,4 +116,26 @@ public boolean isDisableTlsHostnameVerification() {
...
+    }
+
+    public void setMaxTokenExpirySeconds(int maxTokenExpirySeconds) {
+        this.maxTokenExpirySeconds = maxTokenExpirySeconds;
+    }
+
+    @Description("Configure if access token should not be treated as JWT. Set this to `true` if authorization " +
+            "server returns opaque tokens. Defaults to `false`.")
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    public boolean isNotJwt() {
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java]: I wonder if we have some better name for this. Maybe just `jwt: true` / `jwt: false`? Ot `jwtFormat`? WDYT @tombentley?

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java [hunk 2]
```diff
@@ -40,6 +40,8 @@
     private int jwksExpirySeconds;
     private String introspectionEndpointUri;
     private String userNameClaim;
+    private boolean skipTypeCheck;
+    private boolean notJwt;
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]: Should we set it to `false` as default? Or is it not needed?

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java [hunk 3]
```diff
@@ -138,6 +140,28 @@ public void setUserNameClaim(String userNameClaim) {
...
+    }
+
+    public void setSkipTypeCheck(boolean skipTypeCheck) {
+        this.skipTypeCheck = skipTypeCheck;
+    }
+
+    @Description("Configure if access token should not be treated as JWT. Set this to `true` if authorization " +
+            "server returns opaque tokens. Defaults to `false`.")
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    public boolean isNotJwt() {
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]: Same as above .... I wonder about the name.

File: api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java [hunk 4]
```diff
@@ -114,4 +116,26 @@ public boolean isDisableTlsHostnameVerification() {
     public void setDisableTlsHostnameVerification(boolean disableTlsHostnameVerification) {
         this.disableTlsHostnameVerification = disableTlsHostnameVerification;
     }
+
+    @Description("Set or limit time-to-live of the access tokens to the specified number of seconds. " +
+            "Set this if authorization server returns opaque tokens.")
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java]:
```suggestion
            "This should be set if the authorization server returns opaque tokens.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java [hunk 5]
```diff
@@ -114,4 +116,26 @@ public boolean isDisableTlsHostnameVerification() {
...
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    public int getMaxTokenExpirySeconds() {
+        return maxTokenExpirySeconds;
+    }
+
+    public void setMaxTokenExpirySeconds(int maxTokenExpirySeconds) {
+        this.maxTokenExpirySeconds = maxTokenExpirySeconds;
+    }
+
+    @Description("Configure if access token should not be treated as JWT. Set this to `true` if authorization " +
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java]:
```suggestion
    @Description("Configure whether access token should not be treated as JWT. This should be set to `true` if the authorization " +
```

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java [hunk 6]
```diff
@@ -138,6 +140,28 @@ public void setUserNameClaim(String userNameClaim) {
         this.userNameClaim = userNameClaim;
     }
 
+    @Description("Configure if access token type check should be skipped. Set this to `true` if authorization server " +
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]:
```suggestion
    @Description("Configure whether the access token type check should be skipped. This should be set to `true` if the authorization server " +
```

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java [hunk 7]
```diff
@@ -138,6 +140,28 @@ public void setUserNameClaim(String userNameClaim) {
...
+            "does not include 'typ' claim in JWT token. Defaults to `false`.")
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]: I think typ is not a typo, but I just want to check.

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java [hunk 8]
```diff
@@ -138,6 +140,28 @@ public void setUserNameClaim(String userNameClaim) {
...
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    public boolean isSkipTypeCheck() {
+        return skipTypeCheck;
+    }
+
+    public void setSkipTypeCheck(boolean skipTypeCheck) {
+        this.skipTypeCheck = skipTypeCheck;
+    }
+
+    @Description("Configure if access token should not be treated as JWT. Set this to `true` if authorization " +
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]:
```suggestion
    @Description("Configure whether access tokens should not be treated as JWT. This should be set to `true` if the authorization " +
```

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java [hunk 9]
```diff
@@ -1673,6 +1673,26 @@ private static void validateOauth(KafkaListenerAuthenticationOAuth oAuth, String
             log.error("{}: The refresh interval has to be at least 60 seconds shorter then the expiry interval specified in `jwksExpirySeconds`", listener);
             throw new InvalidResourceException(listener + ": The refresh interval has to be at least 60 seconds shorter then the expiry interval specified in `jwksExpirySeconds`");
         }
+
+        if (oAuth.isNotJwt()) {
+            if (oAuth.getJwksEndpointUri() != null) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: Should we also check these properties for emptiness, just to be safe?

File: api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java (hunk 1 above)
Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java]: Yeah, I wondered about this too. It's always a bit weird having a negative property and having to set notFoo=true. How about `accessTokenIsJwt` defaulting to true?

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java [hunk 10]
```diff
@@ -138,6 +140,28 @@ public void setUserNameClaim(String userNameClaim) {
         this.userNameClaim = userNameClaim;
     }
 
+    @Description("Configure if access token type check should be skipped. Set this to `true` if authorization server " +
+            "does not include 'typ' claim in JWT token. Defaults to `false`.")
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    public boolean isSkipTypeCheck() {
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]: Would `checkAccessTokenType` (defaulting to true) be a better name?

File: api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java (hunk 1 above)
Comment (scholzj) [reply api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java]: `accessTokenIsJwt` sounds good to me 👍

Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java]: Yeah tbh I don't like "negative" properties as well :-) `accessTokenIsJwt` sound fine!

Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java]: Very well. Changed to `accessTokenIsJwt`. The same can be changed in strimzi-kafka-oauth library for its next release and the necessary modifications made here behind the scenes.

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java (hunk 2 above)
Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]: The default of the 'accessTokenIsJwt' is now `true`. No need for it to be explicitly set.

File: api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java (hunk 4 above)
Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java]: Fixed.

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java (hunk 6 above)
Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]: Fixed.

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java (hunk 8 above)
Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]: Fixed.

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java (hunk 10 above)
Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]: Sure, why not. Changed the name to `checkAccessTokenType`.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java (hunk 9 above)
Comment (mstruk) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: Not following here. The `accessTokenIsJwt` is `boolean` - can't be empty. `jwksEndpointUri` is null unless set in which case it won't be null.

File: api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java (hunk 7 above)
Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java]: Correct. Not a typo.

File: api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java (hunk 5 above)
Comment (mstruk) [reply api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java]: Fixed.

**Comment by mstruk** [comment]
We may now want to rename the title of this PR to:

Add accessTokenIsJwt, maxTokenExpirySeconds, and checkAccessTokenType to CRDs

**Comment by scholzj** [comment]
@strimzi-ci run tests

**Comment by scholzj** [comment]
@strimzi-ci run tests profile=acceptance testcase=io.strimzi.systemtest.KafkaST#testLoadBalancerTls,io.strimzi.systemtest.ConnectST#testKafkaConnectWithFileSinkPlugin

**Pull request merged**